teachertool/src/services/indexedDbService.ts

```diff
@@ -0,0 +1,128 @@
+        if (this.db) return;
+        this.db = await openDB(teacherToolDbName, dbVersion, {
+            upgrade(db) {
+                db.createObjectStore(rubricsStoreName, { keyPath: "name" });
+                db.createObjectStore(metadataStoreName, { keyPath: "key" });
+            },
+        });
+    }
+
+    private async getAsync<T>(storeName: string, key: string): Promise<T | undefined> {
+        if (!this.db) {
+            throw new Error("IndexedDb not initialized.");
+        }
+
+        try {
+            return await this.db.get(storeName, key);
+        } catch (e) {
+            // Not recording key, as it could contain user-input with sensitive information.
+            logError(ErrorCode.unableToGetIndexedDbRecord, e);
+        }
+    }
+
+    private async setAsync<T>(storeName: string, value: T): Promise<void> {
+        if (!this.db) {
+            throw new Error("IndexedDb not initialized.");
+        }
+
+        try {
+            await this.db.put(storeName, value);
+        } catch (e) {
+            // Not recording key, as it could contain user-input with sensitive information.
+            logError(ErrorCode.unableToSetIndexedDbRecord, e);
+        }
+    }
+
+    private async deleteAsync(storeName: string, key: string): Promise<void> {
+        if (!this.db) {
+            throw new Error("IndexedDb not initialized.");
+        }
+        try {
+            await this.db.delete(storeName, key);
+        } catch (e) {
+            // Not recording key, as it could contain user-input with sensitive information.
+            logError(ErrorCode.unableToDeleteIndexedDbRecord, e);
+        }
+    }
+
+    private async getMetadataEntryAsync(key: string): Promise<MetadataEntry | undefined> {
+        return this.getAsync<MetadataEntry>(metadataStoreName, key);
+    }
+
+    private async setMetadataEntryAsync(key: string, value: any): Promise<void> {
+        return this.setAsync<MetadataEntry>(metadataStoreName, { key, value });
+    }
+
+    private async deleteMetadataEntryAsync(key: string): Promise<void> {
+        return this.deleteAsync(metadataStoreName, key);
+    }
+
+    public async getLastActiveRubricNameAsync(): Promise<string | undefined> {
+        const metadataEntry = await this.getMetadataEntryAsync(metadataKeys.lastActiveRubricKey);
+        return metadataEntry?.value;
+    }
+
+    public saveLastActiveRubricNameAsync(name: string): Promise<void> {
+        return this.setMetadataEntryAsync(metadataKeys.lastActiveRubricKey, name);
+    }
+
+    public getRubric(name: string): Promise<Rubric | undefined> {
+        return this.getAsync<Rubric>(rubricsStoreName, name);
+    }
+
+    public saveRubric(rubric: Rubric): Promise<void> {
+        return this.setAsync(rubricsStoreName, rubric);
+    }
+
+    public deleteRubric(name: string): Promise<void> {
+        return this.deleteAsync(rubricsStoreName, name);
+    }
+}
+
+const db = new TeacherToolDb();
+
+let initializeAsync = async () => {
```

Comment:
This code actually doesn't work as intended. Due to the async, the function can be called multiple times before the reassignment happens. This is my bug :). I introduced it in Kiosk. One fix would be to have a global promise that does the db init and then resolves. getLastActiveRubricAsync, etc. would wait on that.

Reply:
You could create a global promise that is the only source of the db instance. API calls have to await it to get the db instance. This would ensure no calls can be done on `db` before it's initialized.